### PR TITLE
Basemaps: remove custom baselayer context menu

### PR DESCRIPTION
**Pull request by matallo**
CR @nobuti 

- added remove basemap context menu
- tests not added yet, to confirm implementation, will be added

Take into account while removing the custom basemap, it will still remain as a baselayer (they are different things) until the baselayer changes. This behaviour is consistent with current editor.

remove

@CartoDB/design I just made the implementation, but expect the context menu cue (in the bottom right of the mosaics) to change accordingly to the last designs, feel free to followup discussion here and I'll update then

also, now this context menu opens from bottom left corner, as opening from bottom right corner will make the menu to crop in the left border of the screen

/cc @xavijam

### CODE REVIEW

File: app/assets/stylesheets/editor-3/mosaic.css.scss [hunk 1]
```diff
@@ -71,3 +78,29 @@
...
   line-height: 40px;
 }
+.Mosaic-remove {
+  position: absolute;
+  right: 0;
+  bottom: 0;
+  width: 14px;
+  height: 14px;
+  padding: 0;
+  transition: opacity 0.1s;
```
Comment (nobuti) [inline app/assets/stylesheets/editor-3/mosaic.css.scss]: mixin here

File: lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-remove-view.js [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+    var triggerElementID = 'context-menu-trigger-' + this.model.cid;
+    this.$('.js-Mosaic-remove').attr('id', triggerElementID);
+    this._menuView = new ContextMenuView({
+      collection: menuItems,
+      triggerElementID: triggerElementID,
+      position: position
+    });
+
+    menuItems.bind('change:selected', function (menuItem) {
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-remove-view.js]: I feel there would be an unbind somewhere else. I'd define `menuItems` as a property of `this`:
```
initialize: function (opts) {
...
this._menuItems = new CustomListCollection([{
    label: 'Remove baselayer',
    val: 'remove-baselayer',
    destructive: true
}]);

this._menuItems.on('change:selected', function (menuItem) {
...
});
this.add_related_model(this._menuItems);
},
```

File: lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-remove-view.js [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+  _hideContextMenu: function () {
+    this._menuView.remove();
+    this.removeView(this._menuView);
+    delete this._menuView;
+  },
+
+  _showContextMenu: function (position) {
+    var menuItems = new CustomListCollection([{
+      label: 'Remove baselayer',
```
Comment (nobuti) [inline lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-remove-view.js]: Locale here

File: lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-remove-view.js [hunk 4]
```diff
@@ -0,0 +1,82 @@
+var CoreView = require('backbone/core-view');
+var template = require('./basemap-mosaic-remove.tpl');
+var CustomListCollection = require('../../../components/custom-list/custom-list-collection');
+var ContextMenuView = require('../../../components/context-menu/context-menu-view');
+
+module.exports = CoreView.extend({
+
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-remove-view.js]: You can avoid template with this.
…
tagName: 'span',
className: 'whatever',
…

Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-remove-view.js]: It's @xavijam

File: lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-remove-view.js [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+  },
+
+  _hideContextMenu: function () {
+    this._menuView.remove();
+    this.removeView(this._menuView);
+    delete this._menuView;
+  },
+
+  _showContextMenu: function (position) {
+    var menuItems = new CustomListCollection([{
```
Comment (matallo) [inline lib/assets/javascripts/cartodb3/editor/layers/basemap-content-views/basemap-mosaic-remove-view.js]: Check that there is already a Klass that manages the context-menu creation.

**Comment by xavijam** [comment]
ping!

**Comment by matallo** [comment]
I'm making the changes but kept with the context menu boilerplate, as using the factory wasn't that straightforward, some of the drawbacks I found (see attached animation):

- I'd still need to use a template for the `.js-` child classes
- need to stop propagation of events
- the styles are not flexible enough

basemapsremove

so for this time, I'm using the views and collections but that could be something to take into account for future improvements over that component

File: app/assets/stylesheets/editor-3/mosaic.css.scss (hunk 1 above)
Comment (matallo) [inline app/assets/stylesheets/editor-3/mosaic.css.scss]: true dat

I want to move to a postprocessor for this stuff :(

**Comment by matallo** [comment]
pong!

screen shot 2016-10-05 at 13 16 07

styles updated, changes applied /cc @nobuti @arianaescobar

**Pull request merged**